java/core/src/main/java/com/google/protobuf/Utf8.java

```diff
@@ -610,6 +634,98 @@ private static int partialIsValidUtf8(final ByteBuffer buffer, int index, final
     }
 
     /**
+     * Decodes the given byte array slice into a {@link String}.
+     *
+     * @throws InvalidProtocolBufferException if the byte array slice is not valid UTF-8.
+     */
+    abstract String decodeUtf8(byte[] bytes, int index, int size)
+        throws InvalidProtocolBufferException;
+
+    /**
+     * Decodes the given portion of the {@link ByteBuffer} into a {@link String}.
+     *
+     * @throws InvalidProtocolBufferException if the portion of the buffer is not valid UTF-8.
+     */
+    final String decodeUtf8(ByteBuffer buffer, int index, int size)
+        throws InvalidProtocolBufferException {
+      if (buffer.hasArray()) {
+        final int offset = buffer.arrayOffset();
+        return decodeUtf8(buffer.array(), offset + index, size);
+      } else if (buffer.isDirect()) {
+        return decodeUtf8Direct(buffer, index, size);
+      }
+      return decodeUtf8Default(buffer, index, size);
+    }
+
+    /**
+     * Decodes direct {@link ByteBuffer} instances into {@link String}.
+     */
+    abstract String decodeUtf8Direct(ByteBuffer buffer, int index, int size)
+        throws InvalidProtocolBufferException;
+
+    /**
+     * Decodes {@link ByteBuffer} instances using the {@link ByteBuffer} API rather than
+     * potentially faster approaches.
+     */
+    final String decodeUtf8Default(ByteBuffer buffer, int index, int size)
+        throws InvalidProtocolBufferException {
+      if ((index | size | buffer.limit() - index - size) < 0) {
+        throw new ArrayIndexOutOfBoundsException(
+            String.format("buffer limit=%d, index=%d, limit=%d", buffer.limit(), index, size));
+      }
+
+      int offset = index;
+      final int limit = offset + size;
+
+      char[] resultArr = new char[size];
+      int resultPos = 0;
+
+      // Optimize for 100% ASCII (Hotspot loves small simple top-level loops like this).
+      // This simple loop stops when we encounter a byte >= 0x80 (i.e. non-ASCII).
+      while (offset < limit) {
+        byte b = buffer.get(offset);
+        if (b >= 0) {
+          offset++;
+          resultArr[resultPos++] = (char) b;
+        } else {
+          break;
+        }
+      }
+
+      while (offset < limit) {
+        byte byte1 = buffer.get(offset++);
+        if (byte1 >= 0) {
+          DecodeUtil.handleOneByte(byte1, resultArr, resultPos++);
+        } else if (byte1 < (byte) 0xE0) {
+          if (offset >= limit || byte1 < (byte) 0xC2) {
+            throw InvalidProtocolBufferException.invalidUtf8();
+          }
+          byte byte2 = buffer.get(offset++);
+          DecodeUtil.handleTwoBytes(byte1, byte2, resultArr, resultPos++);
+        } else if (byte1 < (byte) 0xF0) {
+          if (offset >= limit - 1) {
+            throw InvalidProtocolBufferException.invalidUtf8();
+          }
+          byte byte2 = buffer.get(offset++);
```

Comment:
It might be better to avoid the store operations and pass `buffer.get(offset++)` direction into the method call.

same comment throughout.